addon/-private/factory-for.js

```diff
@@ -2,13 +2,22 @@
   Utility to use `ApplicationInstance#factoryFor` if available (2.12+) and
   fallback to private `_lookupFactory` otherwise
 */
+class LegacyFactoryFor {
+  constructor(owner, fullName, options) {
+    this.class = owner._lookupFactory(fullName, options);
+  }
+
+  create(fullName, options) {
+    if (this.class) {
+      return this.class.create(fullName, options);
+    }
+  }
+}
+
 export default function factoryFor(owner, type) {
-  let factory;
   if (owner.factoryFor) {
-    let maybeFactory = owner.factoryFor(type);
-    factory = maybeFactory && maybeFactory.class;
```

Comment:
Source of the issue is returning .class here which does not preserve the owner when instances are instantiated from it. We need to create() from that object that factoryFor returns while still being able to access static properties before instantiation.